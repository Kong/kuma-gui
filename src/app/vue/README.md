---
section: Packages
---
# vue

Automatic `vue` initialization using our projects dependency injection layer.

