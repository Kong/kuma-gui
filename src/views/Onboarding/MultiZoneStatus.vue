<template>
  <OnboardingPage>
    <template #header>
      <OnboardingHeading title="Adding a new zone" />
    </template>
    <template #content>
      <p class="text-center my-4">
        Zone status:
        <span v-if="hasZones" class="text-green-500" data-testid="zone-online">Online</span>
        <span v-else class="text-red-500" data-testid="zone-offline">Offline</span>
      </p>
      <p class="text-center mt-4 mb-10">
        Zone ingress status:
        <span v-if="hasZoneIngresses" class="text-green-500" data-testid="zone-ingress-online">Online</span>
        <span v-else class="text-red-500" data-testid="zone-ingress-offline">Offline</span>
      </p>
      <div v-if="!hasZoneIngresses || !hasZones" class="flex justify-center">
        <Loading />
      </div>
    </template>

    <template #navigation>
      <OnboardingNavigation
        next-step="onboarding-populating-mesh"
        previous-step="onboarding-multi-zone"
        :should-display-next="servicesOnline"
      />
    </template>
  </OnboardingPage>
</template>

<script>
import Kuma from '@/services/kuma'
import Loading from '@/components/Loading'
import OnboardingNavigation from '@/views/Onboarding/components/OnboardingNavigation'
import OnboardingHeading from '@/views/Onboarding/components/OnboardingHeading'
import OnboardingPage from '@/views/Onboarding/components/OnboardingPage'

const LONG_POOLING_INTERVAL = 1000

export default {
  name: 'Multizone',
  components: {
    OnboardingNavigation,
    OnboardingHeading,
    OnboardingPage,
    Loading,
  },
  metaInfo() {
    return {
      title: 'Multizone status',
    }
  },
  data() {
    return {
      hasZones: false,
      hasZoneIngresses: false,
      zoneTimeout: null,
      zoneIngressTimeout: null,
    }
  },
  computed: {
    servicesOnline() {
      return this.hasZoneIngresses && this.hasZones
    },
  },

  created() {
    this.getZones()
    this.getZoneIngresses()
  },

  destroyed() {
    clearTimeout(this.zoneTimeout)
    clearTimeout(this.zoneIngressTimeout)
  },
  methods: {
    async getZones() {
      try {
        const { total } = await Kuma.getZones()

        this.hasZones = total > 0
      } catch (e) {
        console.error(e)
      }

      if (!this.hasZones) {
        this.zoneTimeout = setTimeout(() => this.getZones(), LONG_POOLING_INTERVAL)
      }
    },
    async getZoneIngresses() {
      try {
        const { total } = await Kuma.getAllZoneIngressOverviews()

        this.hasZoneIngresses = total > 0
      } catch (e) {
        console.error(e)
      }

      if (!this.hasZoneIngresses) {
        this.zoneIngressTimeout = setTimeout(() => this.getZoneIngresses(), LONG_POOLING_INTERVAL)
      }
    },
  },
}
</script>
