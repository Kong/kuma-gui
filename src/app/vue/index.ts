import { createApp } from 'vue'
import {
  createRouter,
  createWebHistory,
} from 'vue-router'

import type { ServiceDefinition } from '@/services/utils'
import { token, createInjections } from '@/services/utils'
import type { Component } from 'vue'
import type { Router, RouteRecordRaw, NavigationGuard } from 'vue-router'

type Token = ReturnType<typeof token>

type VueApp = ReturnType<typeof createApp>
export type PluginDefinition = Parameters<VueApp['use']>
export type ComponentDefinition = [string, Component]

const $ = {
  app: token<(App: Component) => Promise<VueApp>>('vue.app'),
  router: token<Router>('vue.router'),

  components: token('vue.components'),
  plugins: token<PluginDefinition[]>('vue.plugins'),
  routes: token<RouteRecordRaw[]>('vue.routes'),
  routesLabel: token<RouteRecordRaw[]>('vue.routes.label'),
  navigationGuards: token<NavigationGuard[]>('vue.routes.navigation.guards'),
}
export const services = (app: Record<string, Token>): ServiceDefinition[] => {
  return [
    [$.app, {
      service: (
        components: ComponentDefinition[],
        plugins: PluginDefinition[],
      ) => {
        return async (App: Component) => {
          const app = createApp(App)

          plugins.forEach(([...args]) => {
            app.use(...args)
          })

          components.forEach(([name, component]: [string, Component]) => {
            app.component(name, component)
          })

          return app
        }
      },
      arguments: [
        $.components,
        $.plugins,
      ],
    }],
    [$.router, {
      service: (
        env: (str: 'KUMA_BASE_PATH') => string,
        routes: RouteRecordRaw[],
        guards: NavigationGuard[],
      ) => {
        const router = createRouter({
          history: createWebHistory(env('KUMA_BASE_PATH')),
          routes,
        })

        guards.forEach((item) => {
          if (typeof item === 'function') {
            router.beforeEach(item)
          }
        })

        return router
      },
      arguments: [
        app.env,
        $.routes,
        $.navigationGuards,
      ],
    }],

    [token('vue.plugins'), {
      service: (router) => {
        return [
          [router],
        ]
      },
      arguments: [
        $.router,
      ],
      labels: [
        $.plugins,
      ],
    }],

    // TODO(jc): Label decoration. Make it so we can decorate labels
    // Temporarily turn the routesLabel label into routes for decoration
    // purposes
    [$.routes, {
      service: (routes: RouteRecordRaw[]) => routes,
      arguments: [
        $.routesLabel,
      ],
    }],
    //

    // TODO(jc): Make it so we don't need to provide a [] empty default for
    // labels
    [token('application.routes.navigation.guards'), {
      service: () => {
        return []
      },
      labels: [
        $.navigationGuards,
      ],
    }],
  ]
}

export const TOKENS = $
export const [
  useRouter,
] = createInjections(
  TOKENS.router,
)
