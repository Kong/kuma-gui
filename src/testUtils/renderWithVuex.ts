import { render, RenderOptions, RenderResult } from '@testing-library/vue'
import merge from 'lodash/merge'
import { StoreOptions } from 'vuex'
import Vue, { ComponentOptions } from 'vue'
import Store from '@/store'

function renderWithVuex(
  Component: typeof Vue | ComponentOptions<Vue>,
  options: RenderOptions<Vue> = { store: {}, routes: [] },
): RenderResult {
  const store: StoreOptions<any> = Store()

  // Render the component and merge the original store and the custom one
  // provided as a parameter. This way, we can alter some behaviors of the
  // initial implementation.

  return render(Component, { ...options, store: merge(store, options.store), routes: options.routes })
}

export default renderWithVuex
