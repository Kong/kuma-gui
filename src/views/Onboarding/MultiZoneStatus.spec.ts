import renderWithVuex from '@/testUtils/renderWithVuex'
import { screen } from '@testing-library/vue'

import Kuma from '@/services/kuma'
import MultiZoneStatus from './MultiZoneStatus.vue'

describe('MultiZoneStatus.vue', () => {
  it('renders snapshot', () => {
    const { container } = renderWithVuex(MultiZoneStatus)

    expect(container).toMatchSnapshot()
  })

  it('detects resources on call and allow to proceed', async () => {
    renderWithVuex(MultiZoneStatus, {
      routes: [],
    })

    expect(await screen.findByTestId('zone-online')).toBeInTheDocument()
    expect(await screen.findByTestId('zone-ingress-online')).toBeInTheDocument()
    expect(screen.queryByText(/Next/)).toBeInTheDocument()
    expect(screen.queryByTestId('loading')).not.toBeInTheDocument()
  })

  it('refetch resources if any not available', async () => {
    jest.useFakeTimers()
    jest
      .spyOn(Kuma, 'getAllZoneIngressOverviews')
      .mockReturnValueOnce({
        // @ts-ignore
        total: 0,
      })
      .mockReturnValueOnce({
        // @ts-ignore
        total: 1,
      })

    renderWithVuex(MultiZoneStatus, {
      routes: [],
    })

    expect(await screen.findByTestId('zone-online')).toBeInTheDocument()
    expect(screen.queryByTestId('loading')).toBeInTheDocument()
    expect(screen.queryByTestId('zone-ingress-offline')).toBeInTheDocument()

    jest.runAllTimers()

    expect(Kuma.getAllZoneIngressOverviews).toHaveBeenCalledTimes(2)

    expect(await screen.findByTestId('zone-ingress-online')).toBeInTheDocument()
  })
})
