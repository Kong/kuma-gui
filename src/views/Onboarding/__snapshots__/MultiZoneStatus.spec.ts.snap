// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MultiZoneStatus.vue renders snapshot 1`] = `
<div>
  <div>
    <div
      class="onboarding-container"
    >
      <div
        class="onboarding-container__header"
      >
        <div
          class="relative"
        >
          <h1
            class="onboarding-title"
          >
            
    Adding a new zone
  
          </h1>
           
          <!---->
        </div>
      </div>
       
      <div
        class="onboarding-container__content"
      >
        <div
          class="w-full"
        >
          <p
            class="text-center my-4"
          >
            
      Zone status:
      
            <span
              class="text-red-500"
              data-testid="zone-offline"
            >
              Offline
            </span>
          </p>
           
          <p
            class="text-center mt-4 mb-10"
          >
            
      Zone ingress status:
      
            <span
              class="text-red-500"
              data-testid="zone-ingress-offline"
            >
              Offline
            </span>
          </p>
           
          <div
            class="flex justify-center"
          >
            <div
              class="loading"
              data-testid="loading"
            >
              <div />
               
              <div />
               
              <div />
               
              <div />
               
              <div />
            </div>
          </div>
        </div>
      </div>
       
      <div
        class="mt-4 flex items-center flex-col sm:flex-row justify-between"
      >
        <a
          aria-current="page"
          class="k-button navigation-button navigation-button--back router-link-exact-active router-link-active primary"
          data-v-38fa8ecf=""
          href="#/"
          type="button"
        >
          
    Back
  
        </a>
         
        <!---->
      </div>
    </div>
     
    <div
      class="background-image"
    />
  </div>
</div>
`;
